vortex-file/src/scan/mod.rs

```diff
@@ -0,0 +1,286 @@
+            split_by: SplitBy::Layout,
+            canonicalize: false,
+            // How many row splits to make progress on concurrently (not necessarily in parallel,
+            // that is decided by the TaskExecutor).
+            concurrency: 16,
+            metrics,
+        }
+    }
+
+    pub fn with_filter(mut self, filter: ExprRef) -> Self {
+        self.filter = Some(filter);
+        self
+    }
+
+    pub fn with_some_filter(mut self, filter: Option<ExprRef>) -> Self {
+        self.filter = filter;
+        self
+    }
+
+    pub fn with_projection(mut self, projection: ExprRef) -> Self {
+        self.projection = projection;
+        self
+    }
+
+    pub fn with_row_indices(mut self, row_indices: Buffer<u64>) -> Self {
+        self.row_indices = Some(row_indices);
+        self
+    }
+
+    pub fn with_some_row_indices(mut self, row_indices: Option<Buffer<u64>>) -> Self {
+        self.row_indices = row_indices;
+        self
+    }
+
+    pub fn with_split_by(mut self, split_by: SplitBy) -> Self {
+        self.split_by = split_by;
+        self
+    }
+
+    /// Set whether the scan should canonicalize the output.
+    pub fn with_canonicalize(mut self, canonicalize: bool) -> Self {
+        self.canonicalize = canonicalize;
+        self
+    }
+
+    /// The number of row splits to make progress on concurrently, must be greater than 0.
+    pub fn with_concurrency(mut self, concurrency: usize) -> Self {
+        assert!(concurrency > 0);
+        self.concurrency = concurrency;
+        self
+    }
+
+    pub fn with_task_executor(mut self, task_executor: TaskExecutor) -> Self {
+        self.task_executor = Some(task_executor);
+        self
+    }
+
+    pub fn with_metrics(mut self, metrics: VortexMetrics) -> Self {
+        self.metrics = metrics;
+        self
+    }
+
+    #[allow(clippy::unused_enumerate_index)]
+    pub fn build(self) -> VortexResult<impl ArrayStream + 'static> {
+        // Spin up the root layout reader, and wrap it in a FilterLayoutReader to perform
+        // conjunction splitting.
+        let layout_reader = Arc::new(FilterLayoutReader::new(self.vxf.layout_reader()?));
+
+        // Normalize and simplify the expressions.
+        let projection = simplify_typed(self.projection.clone(), layout_reader.dtype())?;
+        let filter = self
+            .filter
+            .clone()
+            .map(|f| simplify_typed(f, layout_reader.dtype()))
+            .transpose()?;
+
+        // Construct field masks and compute the row splits of the scan.
+        let (filter_mask, projection_mask) =
+            filter_and_projection_masks(&projection, filter.as_ref(), layout_reader.dtype())?;
+        let field_mask: Vec<_> = filter_mask
+            .iter()
+            .cloned()
+            .chain(projection_mask.iter().cloned())
+            .collect();
+        let splits = self.split_by.splits(layout_reader.layout(), &field_mask)?;
+        let row_indices = self.row_indices.clone();
```

Comment:
after this, maybe we do a row indices layout and wrap the filter layout if row_indices are present